### PR TITLE
Change python interface for key_by, select/annotate/drop

### PR DESCRIPTION
**Pull request by catoverdrive**
currently just for the rows of MatrixTable, but I'd appreciate some feedback on what I've got so far. I think I'm probably just going to make all the interface changes on this branch/in this PR so they're all in one place before they go in?

from:
http://dev.hail.is/t/proposed-changes-to-the-python-select-annotate-drop-interface-for-keys/88

### CODE REVIEW

**Comment by catoverdrive** [comment]
@tpoterba I think I got the rest of them now, too. (Table.row, MatrixTable.col) I kind of want a better way to talk about the full row struct of a Table, or at least the dtype of that. What do you think?

**Comment by tpoterba** [comment]
getting back to this.

File: python/hail/matrixtable.py [hunk 1]
```diff
@@ -657,78 +647,88 @@ def entry(self):
...
 
+        See :meth:`.MatrixTable.key_rows_by` for more examples.
+
         Parameters
         ----------
-        keys : varargs of :obj:`str`
-            Column fields to key by.
+        keys : varargs of :obj:`str` or :class:`.Expression`.
+            Row fields to key by.
+        named_keys : varargs of :class:`.Expression`.
```
Comment (tpoterba) [inline python/hail/matrixtable.py]: keyword args

File: python/hail/table.py [hunk 2]
```diff
@@ -372,11 +372,25 @@ def count(self):
...
 
-    @typecheck_method(caller=str, s=expr_struct())
-    def _select(self, caller, s):
-        base, cleanup = self._process_joins(s)
-        analyze(caller, s, self._row_indices)
-        return cleanup(Table(base._jt.select(s._ast.to_hql())))
+    @typecheck_method(caller=str, key_struct=nullable(expr_struct()), value_struct=nullable(expr_struct()))
+    def _select(self, caller, key_struct=None, value_struct=None):
+        if key_struct is None:
+            assert(value_struct is not None)
```
Comment (tpoterba) [inline python/hail/table.py]: unnecessary parens

File: src/test/scala/is/hail/testUtils/RichTable.scala [hunk 3]
```diff
@@ -88,44 +88,11 @@ class RichTable(ht: Table) {
...
-
-      val inserterBuilder = new ArrayBuilder[Inserter]()
-
-      val finalSignature = (paths, types).zipped.foldLeft(ht.signature) { case (vs, (ids, sig)) =>
-        val (s: TStruct, i) = vs.insert(sig, ids)
-        inserterBuilder += i
-        s
-      }
+  def keyByExpr(exprs: (String, String)*): Table =
+    ht.select(s"annotate(row, {${ exprs.map { case (n, e) => s"`$n`: $e" }.mkString(",") }})", Some(exprs.map(_._1).toIndexedSeq))
```
Comment (tpoterba) [inline src/test/scala/is/hail/testUtils/RichTable.scala]: can't just backtick it like this -- the field name could have a backtick in it. Use prettyIdentifier (I think).

File: src/test/scala/is/hail/testUtils/RichTable.scala [hunk 4]
```diff
@@ -88,44 +88,11 @@ class RichTable(ht: Table) {
...
-          }
-      }
-
-      ht.copy(rdd = ht.rdd.map(annotF), signature = finalSignature, key = ht.key)
-    } else {
-      val newIR = InsertFields(Ref("row"), paths.zip(irs))
-      new Table(ht.hc, TableMapRows(ht.tir, newIR))
-    }
-  }
+  def annotate(exprs: (String, String)*): Table =
```
Comment (tpoterba) [inline src/test/scala/is/hail/testUtils/RichTable.scala]: same as above

File: src/main/scala/is/hail/variant/MatrixTable.scala [hunk 5]
```diff
@@ -1003,47 +1005,24 @@ class MatrixTable(val hc: HailContext, val ast: MatrixIR) {
       colValues = colValues.copy(newColValues, TArray(newColType)))
   }
 
-  def selectRows(expr: String): MatrixTable = {
+  def selectRows(expr: String, newKey: java.util.ArrayList[java.util.ArrayList[String]]): MatrixTable = {
+    assert(newKey == null || newKey.size() == 2)
```
Comment (tpoterba) [inline src/main/scala/is/hail/variant/MatrixTable.scala]: `Option(newKey).forall(_.size == 2)`

File: src/main/scala/is/hail/table/Table.scala [hunk 6]
```diff
@@ -918,9 +919,8 @@ class Table(val hc: HailContext, val tir: TableIR) {
 
     val relatedNodesToKeep = this.maximalIndependentSet(iExpr, jExpr, maybeTieBreaker).toSet
 
-    val nodes = this.select(s"{node : [$iExpr, $jExpr] }")
+    val nodes = this.select(s"{node : [$iExpr, $jExpr] }", Some(IndexedSeq("node")))
```
Comment (tpoterba) [inline src/main/scala/is/hail/table/Table.scala]: use FastIndexedSeq() or something now. The thing Cotton added.

File: src/main/scala/is/hail/table/Table.scala [hunk 7]
```diff
@@ -422,14 +422,18 @@ class Table(val hc: HailContext, val tir: TableIR) {
     copy2(key = key.toArray[String])
   }
 
-  def select(expr: String): Table = {
+  def select(expr: String, newKey: java.util.ArrayList[String]): Table =
+    select(expr, Option(newKey).map(_.asScala.toIndexedSeq))
+
+  def select(expr: String, newKey: Option[IndexedSeq[String]]): Table = {
+    info(s"Table.select: $expr")
```
Comment (tpoterba) [inline src/main/scala/is/hail/table/Table.scala]: remove

File: src/main/scala/is/hail/table/Table.scala [hunk 8]
```diff
@@ -422,14 +422,18 @@ class Table(val hc: HailContext, val tir: TableIR) {
     copy2(key = key.toArray[String])
   }
 
-  def select(expr: String): Table = {
+  def select(expr: String, newKey: java.util.ArrayList[String]): Table =
+    select(expr, Option(newKey).map(_.asScala.toIndexedSeq))
```
Comment (tpoterba) [inline src/main/scala/is/hail/table/Table.scala]: toFastIndexedSeq

File: python/hail/utils/misc.py [hunk 9]
```diff
@@ -260,12 +260,53 @@ def get_nice_field_error(obj, item):
...
     for k, v in Counter(fields).items():
         if v > 1:
             from hail.expr.expressions import ExpressionException
-            raise ExpressionException("selection would produce duplicate field {}".format(repr(k)))
+            raise ExpressionException("selection would produce duplicate field '{}'".format(repr(k)))
+
+def check_keys(name, indices):
+    from hail.expr.expressions import ExpressionException
+    if name in set(indices.key):
+        msg = "cannot overwrite key field '{}' with annotate or select; use key_by to modify keys.".format(name)
```
Comment (tpoterba) [inline python/hail/utils/misc.py]: use just `{}` and `repr(name)`

This will escape unprintable characters appropriately

File: python/hail/utils/misc.py [hunk 10]
```diff
@@ -260,12 +260,53 @@ def get_nice_field_error(obj, item):
...
+    from hail.expr.expressions import to_expr, ExpressionException, TopLevelReference, Select
+    exprs = [to_expr(e) if not isinstance(e, str) else indices.source[e] for e in exprs]
+    named_exprs = {k: to_expr(v) for k, v in named_exprs.items()}
+    assignments = OrderedDict()
+
+    for e in exprs:
+        if not e._indices == indices:
+            raise ExpressionException("method '{}' parameter 'exprs' expects {}-indexed fields,"
+                                  " found indices {}".format(caller, list(indices.axes), list(e._indices.axes)))
+        if e._ast.search(lambda ast: not isinstance(ast, TopLevelReference) and not isinstance(ast, Select)):
```
Comment (tpoterba) [inline python/hail/utils/misc.py]: I don't like this. I know it's used elsewhere, but ick.

Maybe OK for now, but how should we fix? Store a dict of all nested fields in the object? That might do it.

File: python/hail/utils/misc.py [hunk 11]
```diff
@@ -260,12 +260,53 @@ def get_nice_field_error(obj, item):
...
+    return assignments
+
+def get_annotate_exprs(caller, named_exprs, indices):
+    from hail.expr.expressions import to_expr, ExpressionException
+    named_exprs = {k: to_expr(v) for k, v in named_exprs.items()}
+    for k, v in named_exprs.items():
+        check_keys(k, indices)
+        if k in indices.key.keys():
+            raise ExpressionException("'{}' cannot overwrite key field: '{}'"
+                                      .format(caller, k))
```
Comment (tpoterba) [inline python/hail/utils/misc.py]: use repr(k) and remove the single quotes in the string

File: python/hail/utils/misc.py [hunk 12]
```diff
@@ -260,12 +260,53 @@ def get_nice_field_error(obj, item):
 def check_collisions(fields, name, indices):
     from hail.expr.expressions import ExpressionException
     if name in fields and not fields[name]._indices == indices:
-        msg = 'name collision with field indexed by {}: {}'.format(list(fields[name]._indices.axes), name)
+        msg = "name collision with field indexed by {}: '{}'".format(list(fields[name]._indices.axes), name)
```
Comment (tpoterba) [inline python/hail/utils/misc.py]: use repr(name) instead of adding quotes

**Comment by tpoterba** [comment]
I think this is a really good change. I'm going to find a time to talk to Konrad to see if he has ideas as well.

File: python/hail/table.py (hunk 2 above)
Comment (catoverdrive) [reply python/hail/table.py]: how does that work? just `assert value_struct is not None`?

Comment (tpoterba) [reply python/hail/table.py]: yep

File: python/hail/utils/misc.py (hunk 10 above)
Comment (catoverdrive) [reply python/hail/utils/misc.py]: you mean like _inverse_fields, but nested? I could do that.

Comment (catoverdrive) [reply python/hail/utils/misc.py]: would that maybe get super large, though?

Comment (tpoterba) [reply python/hail/utils/misc.py]: you're right. How about we just abstract this to a function in the AST module, `is_nested_reference`?

Comment (catoverdrive) [reply python/hail/utils/misc.py]: ok. let me know if what I did is too silly.

**Comment by tpoterba** [comment]
Sorry for putting this off, you're probably going to have a hellish rebase. I'll get to this before the end of the week.

File: python/hail/expr/expr_ast.py [hunk 13]
```diff
@@ -93,6 +101,10 @@ def __init__(self, parent, name):
     def to_hql(self):
         return '{}.{}'.format(self.parent.to_hql(), escape_id(self.name))
 
+    @property
+    def is_nested_field(self):
+        return self.parent.is_nested_field
```
Comment (tpoterba) [inline python/hail/expr/expr_ast.py]: nice

File: python/hail/expr/expressions/indices.py [hunk 14]
```diff
@@ -28,6 +28,28 @@ def unify(*indices):
 
         return Indices(src, axes)
 
+    @property
+    def key(self):
+        from hail import Table, MatrixTable, struct
```
Comment (tpoterba) [inline python/hail/expr/expressions/indices.py]: prefer `hl.Table` / `hl.struct` to local imports

File: python/hail/expr/expressions/indices.py [hunk 15]
```diff
@@ -28,6 +28,28 @@ def unify(*indices):
...
+        if self.source is None:
+            return None
+        elif isinstance(self.source, Table):
+            if self.source._row_axis in self.axes:
```
Comment (tpoterba) [inline python/hail/expr/expressions/indices.py]: maybe nicer as `if self == self.source._row_indices`? same below

Comment (tpoterba) [reply python/hail/expr/expressions/indices.py]: this cleans up the MT ones even more

File: python/hail/expr/expressions/indices.py [hunk 16]
```diff
@@ -28,6 +28,28 @@ def unify(*indices):
...
+                return self.source.key
+            else:
+                return None
+        else:
+            assert(isinstance(self.source, MatrixTable))
```
Comment (tpoterba) [inline python/hail/expr/expressions/indices.py]: style: `assert isinstance(self.source, MT)`
(don't need parens on assert)

File: python/hail/expr/expressions/typed_expressions.py [hunk 17]
```diff
@@ -1352,6 +1352,10 @@ def drop(self, *fields):
         return construct_expr(StructOp('drop', self._ast, *to_drop), result_type,
                               self._indices, self._aggregations, self._joins)
 
+    @typecheck_method(other=expr_struct())
+    def concat(self, other):
```
Comment (tpoterba) [inline python/hail/expr/expressions/typed_expressions.py]: I like this, but it needs a short doc and you should decide what to do if there's a duplicate field name in both sides. Currently it'll error with a somewhat nasty message about passing the same keyword arg twice.

File: python/hail/matrixtable.py [hunk 18]
```diff
@@ -614,9 +604,30 @@ def row(self):
         """
         return self._row
 
+    @property
+    def row_value(self):
+        """Returns a struct expression including all non-key row-indexed fields.
+
+        Examples
+        --------
+
```
Comment (tpoterba) [inline python/hail/matrixtable.py]: unnecessary whitespace

File: python/hail/matrixtable.py [hunk 19]
```diff
@@ -614,9 +604,30 @@ def row(self):
...
+        Get the first five non-key row field names:
+
+        .. doctest::
```
Comment (tpoterba) [inline python/hail/matrixtable.py]: I've learned you don't actually need to enclose this in a doctest block. You can delete the directive and bring the two lines below back a tab.

File: python/hail/matrixtable.py [hunk 20]
```diff
@@ -635,6 +646,27 @@ def col(self):
...
+    @property
+    def col_value(self):
+        """Returns a struct expression including all non-key column-indexed fields.
+
+        Examples
+        --------
+
+        Get all non-key column field names:
+
+        .. doctest::
```
Comment (tpoterba) [inline python/hail/matrixtable.py]: same two comments as above: extra whitespace after Example header, and don't need doctest directive

File: python/hail/matrixtable.py [hunk 21]
```diff
@@ -657,78 +689,161 @@ def entry(self):
...
+                raise ValueError("cannot have two row key fields with the same name: '{}'".format(k))
+
+        key_fields = dict(pk_dict, **rest_of_keys_dict)
+
+        return self._select_rows(caller, key_struct=hl.struct(**key_fields), pk_size=len(pk_dict))
+
+    @typecheck_method(partition_key=oneof(str, sequenceof(str)),
+                      keys=oneof(str, Expression),
+                      named_keys=Expression)
+    def partition_rows_by(self, partition_key, *keys, **named_keys) -> 'MatrixTable':
```
Comment (tpoterba) [inline python/hail/matrixtable.py]: I would prefer an interface where this just takes `*partition_key`. I'll explain my thoughts today.

File: python/hail/methods/family_methods.py [hunk 22]
```diff
@@ -231,12 +231,12 @@ def mendel_errors(call, pedigree) -> Tuple[Table, Table, Table, Table]:
                        snp_errors=hl.agg.count_where(hl.is_snp(entries.alleles[0], entries.alleles[1]) &
                                                      hl.is_defined(entries.mendel_code)))
     )
-    table2 = tm.cols()
+    table2 = tm.cols().key_by()
```
Comment (tpoterba) [inline python/hail/methods/family_methods.py]: should this be `key_by(None)`?

File: python/hail/methods/family_methods.py [hunk 23]
```diff
@@ -261,7 +261,7 @@ def mendel_errors(call, pedigree) -> Tuple[Table, Table, Table, Table]:
                               snp_errors=hl.or_else(
                                   hl.agg.array_sum(hl.agg.filter(hl.is_snp(tm.alleles[0], tm.alleles[1]),
                                                                  implicated[tm.mendel_code])),
-                                  [0, 0, 0])).cols()
+                                  [0, 0, 0])).cols().key_by()
```
Comment (tpoterba) [inline python/hail/methods/family_methods.py]: `key_by(None)`?

File: python/hail/methods/family_methods.py [hunk 24]
```diff
@@ -825,14 +824,14 @@ def solve(p_de_novo):
...
-                          '__site_freq',
-                          'proband',
-                          'father',
-                          'mother',
-                          'proband_entry',
-                          'father_entry',
-                          'mother_entry',
-                          'is_female',
-                          **entries.__call).rename({'__site_freq': 'prior'})
+    entries = entries.key_by('locus', 'alleles')
```
Comment (tpoterba) [inline python/hail/methods/family_methods.py]: I think the key is already [locus, alleles, proband]. We can probably leave these as is.

Comment (tpoterba) [reply python/hail/methods/family_methods.py]: Corollary: we should document this better in the method 😄

File: python/hail/methods/statgen.py [hunk 25]
```diff
@@ -1811,7 +1810,7 @@ def pc_relate(call_expr, min_individual_maf, *, k=None, scores_expr=None,
     else:
         raise ValueError("pc_relate: exactly one of 'k' and 'scores_expr' must be set, found neither")
 
-    scores_table = mt.select_cols(__scores=scores_expr).cols()
+    scores_table = mt.select_cols(__scores=scores_expr).cols().key_by().select('__scores')
```
Comment (tpoterba) [inline python/hail/methods/statgen.py]: key_by(None)?

File: python/hail/table.py [hunk 26]
```diff
@@ -403,8 +415,9 @@ def parallelize(cls, rows, schema=None, key=None, n_partitions=None):
                 Env.hc()._jhc, rows.dtype._to_json(rows.value),
                 rows.dtype.element_type._jtype, joption(key), joption(n_partitions)))
 
-    @typecheck_method(keys=nullable(oneof(str, Expression)))
-    def key_by(self, *keys) -> 'Table':
+    @typecheck_method(keys=nullable(oneof(str, Expression)),
+                      named_keys=Expression)
+    def key_by(self, *keys, **named_keys) -> 'Table':
```
Comment (tpoterba) [inline python/hail/table.py]: I think Table.key_by should be the place where we have the most detailed documentation about how the key_by family of methods works. Some of the docs from k_r_b / k_c_b should be moved here

File: python/hail/table.py [hunk 27]
```diff
@@ -1206,15 +1196,15 @@ def joiner(left):
...
+                        left2 = left.select_rows(**{uid: e for uid, e in zip(uids, exprs)})
                         lrt = (left2.rows()
                             .rename({name: u for name, u in zip(left2.row_key, rk_uids)})
                             .key_by(*uids))
 
                         vt = lrt.join(right)
                         # group uids
                         vt = vt.annotate(**{k_uid: Struct(**{u: vt[u] for u in uids})})
-                        vt = vt.drop(*uids)
+                        vt = vt.key_by().drop(*uids)
```
Comment (tpoterba) [inline python/hail/table.py]: should be key_by(None)?

File: python/hail/table.py [hunk 28]
```diff
@@ -1403,7 +1396,7 @@ def collect(self):
         :obj:`list` of :class:`.Struct`
             List of rows.
         """
-        return hl.tarray(self.row.dtype)._from_json(self._jt.collectJSON())
+        return hl.tarray(self._row.dtype)._from_json(self._jt.collectJSON())
```
Comment (tpoterba) [inline python/hail/table.py]: why these changes?

Comment (tpoterba) [reply python/hail/table.py]: we should use public methods when we can. row === _row in this case, right?

File: python/hail/table.py [hunk 29]
```diff
@@ -2129,19 +2127,19 @@ def globals(self) -> 'StructExpression':
...
     def row(self) -> 'StructExpression':
-        """Returns a struct expression including all row-indexed fields.
+        """Returns a struct expression of all row-indexed fields, including keys.
 
         Examples
         --------
         The data type of the row struct:
 
         >>> table1.row.dtype
-        dtype('struct{ID: int32, HT: int32, SEX: str, X: int32, Z: int32, C1: int32, C2: int32, C3: int32}')
```
Comment (tpoterba) [inline python/hail/table.py]: this is an example so I don't much care, but it seems like `Table.row` didn't change so the behavior here shouldn't change?

Comment (tpoterba) [reply python/hail/table.py]: yeah, definitely, given the example in row_value below.

File: python/hail/utils/misc.py (hunk 12 above)
Comment (tpoterba) [reply python/hail/utils/misc.py]: still prefer this change

File: src/main/scala/is/hail/expr/Relational.scala [hunk 30]
```diff
@@ -874,33 +875,13 @@ case class MatrixMapRows(child: MatrixIR, newRow: IR) extends MatrixIR {
   val tAgg = TAggregable(tAggElt, aggSymTab)
 
   val typ: MatrixType = {
-    val newRVRowSet = newRVRow.typ.fieldNames.toSet
-    val newRowKey = child.typ.rowKey.filter(newRVRowSet.contains)
-    val newPartitionKey = child.typ.rowPartitionKey.filter(newRVRowSet.contains)
+    val newRowKey = newKey.map{ case (pk, k) => pk ++ k }.getOrElse(child.typ.rowKey)
```
Comment (tpoterba) [inline src/main/scala/is/hail/expr/Relational.scala]: how do you specify that the new key is actually None?

File: src/main/scala/is/hail/variant/MatrixTable.scala [hunk 31]
```diff
@@ -1002,50 +1005,25 @@ class MatrixTable(val hc: HailContext, val ast: MatrixIR) {
...
+
+  def selectRows(expr: String, newKey: Option[(IndexedSeq[String], IndexedSeq[String])]): MatrixTable = {
     val ec = rowEC
 
     val rowsAST = Parser.parseToAST(expr, ec)
 
     rowsAST.toIR(Some("AGG")) match {
-      case Some(x) if useIR(this.rowAxis, rowsAST) =>
-        new MatrixTable(hc, MatrixMapRows(ast, x))
+      case Some(x)  if useIR(this.rowAxis, rowsAST) =>
```
Comment (tpoterba) [inline src/main/scala/is/hail/variant/MatrixTable.scala]: whitespace change

File: python/hail/table.py (hunk 28 above)
Comment (catoverdrive) [reply python/hail/table.py]: Yeah. This was left over from when row was actually row_value and I forgot to change back.

File: src/main/scala/is/hail/expr/Relational.scala (hunk 30 above)
Comment (catoverdrive) [reply src/main/scala/is/hail/expr/Relational.scala]: This is MatrixMapRows, did you mean for TableMapRows? That's still handled with Table.unkey() and doesn't call into IR yet.

File: python/hail/expr/expressions/typed_expressions.py (hunk 17 above)
Comment (catoverdrive) [reply python/hail/expr/expressions/typed_expressions.py]: i was thinking about this and I think `annotate` does most of what we want this to do?

Comment (catoverdrive) [reply python/hail/expr/expressions/typed_expressions.py]: I've taken it out in favor of an explicit call to `hl.struct(**k, **v)` in the places where it's used. (They're all internal usages where an error would have been a bug). I think `hl.annotate(s1, s2)` does what this would have done for non-overlapping structs, with better-defined behavior for overlapping structs. I could also keep this in and do field-checking explicitly here to throw a better error, if that would still be useful.

File: python/hail/matrixtable.py (hunk 21 above)
Comment (catoverdrive) [reply python/hail/matrixtable.py]: yeah, this was me taking the lazy way out >.> let's talk?
